fgpyo/sam/__init__.py

```diff
@@ -802,38 +804,104 @@ def isize(r1: AlignedSegment, r2: AlignedSegment) -> int:
         return r2_pos - r1_pos
 
 
-def set_pair_info(r1: AlignedSegment, r2: AlignedSegment, proper_pair: bool = True) -> None:
-    """Resets mate pair information between reads in a pair. Requires that both r1
-    and r2 are mapped.  Can be handed reads that already have pairing flags setup or
-    independent R1 and R2 records that are currently flagged as SE reads.
+def sum_of_base_qualities(rec: AlignedSegment, min_quality_score: int = 15) -> int:
+    """Calculate the sum of base qualities score for an alignment record.
+
+    This function is useful for calculating the "mate score" as implemented in samtools fixmate.
+
+    Args:
+        rec: The alignment record to calculate the sum of base qualities from.
+        min_quality_score: The minimum base quality score to use for summation.
+
+    See:
+        [`calc_sum_of_base_qualities()`](https://github.com/samtools/samtools/blob/4f3a7397a1f841020074c0048c503a01a52d5fa2/bam_mate.c#L227-L238)
+        [`MD_MIN_QUALITY`](https://github.com/samtools/samtools/blob/4f3a7397a1f841020074c0048c503a01a52d5fa2/bam_mate.c#L42)
+    """
+    score: int = sum(qual for qual in rec.query_qualities if qual >= min_quality_score)
+    return score
+
+
+def set_mate_info(
+    r1: AlignedSegment,
+    r2: AlignedSegment,
+    is_proper_pair: Callable[[AlignedSegment, AlignedSegment], bool] = is_proper_pair,
+) -> None:
+    """Resets mate pair information between reads in a pair.
 
     Args:
-        r1: read 1
-        r2: read 2 with the same queryname as r1
+        r1: Read 1 (first read in the template).
+        r2: Read 2 with the same query name as r1 (second read in the template).
+        is_proper_pair: A function that takes the two alignments and determines proper pair status.
     """
     if r1.query_name != r2.query_name:
-        raise ValueError("Cannot set pair info on reads with different query names!")
+        raise ValueError("Cannot set mate info on alignments with different query names!")
+
+    for src, dest in [(r1, r2), (r2, r1)]:
+        dest.next_reference_id = src.reference_id
+        dest.next_reference_name = src.reference_name
+        dest.next_reference_start = src.reference_start
+        dest.mate_is_forward = src.is_forward
+        dest.mate_is_mapped = src.is_mapped
+        dest.set_tag("MC", src.cigarstring)
+        dest.set_tag("MQ", src.mapping_quality)
```

Comment:
Does this "just work" when the source read is unmapped?  I.e. does src.cigarstring return None, causing `MC` to get removed?  Or do we end up with `MC:Z:*`?

Reply:
I'll triple check, but it should be the former.

If either field is `None` (such as when the read is unmapped) then the tag is cleared.

Reply:
I confirm these tests pass when `src` is unmapped:

```python
assert not dest.has_tag("MC")

assert dest.has_tag("MQ")
assert dest.get_tag("MQ") == 0

assert dest.has_tag("ms")
assert dest.get_tag("ms") == 3000
```

Since a record's mapping quality is non-optional according to the SAM spec (§ 1.4 Table 1) it makes sense to set the mate's mapping quality to the default value of 0 for an unmapped read, which is what happens. The tag `ms` is the sum of >=15 base qualities as needed by a typical samtools markdup workflow and is not contingent on whether the read is mapped or not, so it also makes sense it would be set for an unmapped read.